front50-gcs/src/main/java/com/netflix/spinnaker/front50/config/GcsConfig.java

```diff
@@ -0,0 +1,174 @@
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.client.RestTemplate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.schedulers.Schedulers;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.concurrent.Executors;
+import java.util.Collections;
+
+@Configuration
+public class GcsConfig {
+  // Refresh every 10 minutes. In practice this either doesnt matter because refreshes are fast enough,
+  // or should be finer tuned. But it seems silly to refresh at a fast rate when changes are generally infrequent.
+  // Actual queries always check to see if the cache is out of date anyway. So this is mostly for the benefit of
+  // keeping other replicas up to date so that last-minute updates have fewer changes in them.
+  private static int APPLICATION_REFRESH_INTERVAL = 10 * 60;
+  private static int PROJECT_REFRESH_INTERVAL = 10 * 60;
+  private static int NOTIFICATION_REFRESH_INTERVAL = 10 * 60;
+  private static int PIPELINE_REFRESH_INTERVAL = 10 * 60;
+  private static int PIPELINE_STRATEGY_REFRESH_INTERVAL = 10 * 60;
+
+  @Value("${spinnaker.gcs.bucket}")
+  private String bucket;
+
+  @Value("${spinnaker.gcs.rootFolder}")
+  private String rootFolder;
+
+  @Value("${providers.google.primaryCredentials.jsonPath}")
+  private String jsonPath;
+
+  @Value("${providers.google.primaryCredentials.project}")
+  private String project;
+
+  @Value("${Implementation-Version:Unknown}")
+  private String applicationVersion;
+
+  private void ensureBucket(Storage storage) throws IOException {
+    Logger log = LoggerFactory.getLogger(getClass());
+    try {
+        Bucket bkt = storage.buckets().get(bucket).execute();
+        boolean has_versioning = bkt.getVersioning().getEnabled();
+        log.info("Bucket versioning is {}.",
+                 has_versioning ? "enabled" : "DISABLED");
+    } catch (HttpResponseException e) {
+      if (e.getStatusCode() == 404) {
+          log.warn("Bucket {} does not exist. Create it in project={}",
+                   bucket, project);
+          Bucket.Versioning versioning = new Bucket.Versioning().setEnabled(true);
+          Bucket spec = new Bucket().setName(bucket).setVersioning(versioning);
+          try {
+              storage.buckets().insert(project, spec).execute();
+          } catch (IOException e2) {
+              log.error("Could not create bucket={} in project={}: {}",
```

Comment:
I was trying to be consistent with S3 for this initial stab. Since you didnt check for the bucket existing (and bucket could be deleted at some point after anyway) I didnt want to prevent a startup and let the error handling on save worry about it. I dont mind putting an exception here.